Update `creating a new blockchain application` - Closes 1647 (#1648)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/build-blockchain/create-blockchain-app.adoc

= Creating a new blockchain application
Mona Bärenfänger <mona@lightcurve.io> Muhammad Talha <muhammad.talha@lightcurve.io>
// Settings
:toc: preamble
:idprefix:
:idseparator: -
:experimental:
:docs_sdk: lisk-sdk::
// URLs
:url_github_guides_bootstrap: https://github.com/LiskHQ/lisk-sdk-examples/tree/development/tutorials/hello
:url_homebrew: https://brew.sh/
:url_nvm: https://github.com/nvm-sh/nvm
// Project URLs
:url_modules_registration: understand-blockchain/sdk/modules-commands.adoc#adding-a-module-to-the-application
:url_guide_config: build-blockchain/configure-app.adoc
:url_guide_module: build-blockchain/create-module.adoc
:url_guide_asset: build-blockchain/create-asset.adoc
:url_guide_plugin: build-blockchain/create-plugin.adoc
:url_references_cli: {docs_sdk}application-cli.adoc#help
:url_references_cli_start: {docs_sdk}application-cli.adoc#start
:url_lisk_commander: {docs_sdk}references/lisk-commander/index.adoc 
:url_lisk_default_modules: {docs_sdk}modules/index.adoc

On this page, you will learn how to create a new blockchain application with the Lisk SDK.

> To get started with the development of a blockchain application most simply and efficiently, it is *recommended* to follow the guide <<bootstrapping-a-new-blockchain-application-with-lisk-commander>>.

Reviewer: Maybe?
```suggestion
To get started with the development of a blockchain application simply and efficiently, it is *recommended* to follow the guide <<bootstrapping-a-new-blockchain-application-with-lisk-commander>>.
```

[source,bash]
----
$ lisk --version
lisk-commander/6.0.0 darwin-arm64 node-v16.18.0
----

==== Toolchain dependencies

Once Lisk Commander is installed, install the following compiler dependencies as well.

[tabs]
====
Ubuntu::
+
--
[source,bash]
----
sudo apt update
sudo apt install -y libtool automake autoconf curl build-essential python2-minimal
----
--
macOS::
+
--
Ensure that {url_homebrew}[Homebrew] is installed.

[source,bash]
----
brew install autoconf automake libtool python2
----
--
====

=== Initializing a new blockchain application

In this example, we want to create a new blockchain application with the name `hello`.
Our blockchain application is divided into two tiers such as a frontend and a blockchain application. 

> Create a dedicated folder `hello/`, which will contain the files for the entire application.

Reviewer: not sure if this is a necessary step, a user can simply create an app anywhere without a dedicated folder so maybe we can mention that to accommodate both frontend and blockchain application we can create a dedicated folder...

   create src/app/modules/.gitkeep
   create src/app/plugins/.gitkeep
   create src/commands/block/get.ts
   create src/commands/blockchain/export.ts
   create src/commands/blockchain/hash.ts
   create src/commands/blockchain/import.ts
   create src/commands/blockchain/reset.ts
   create src/commands/config/create.ts
   create src/commands/config/show.ts
   create src/commands/endpoint/invoke.ts
   create src/commands/generator/disable.ts
   create src/commands/generator/enable.ts
   create src/commands/generator/export.ts
   create src/commands/generator/import.ts
   create src/commands/generator/status.ts
   create src/commands/genesis-block/create.ts
   create src/commands/keys/create.ts
   create src/commands/keys/encrypt.ts
   create src/commands/keys/export.ts
   create src/commands/keys/import.ts
   create src/commands/node/info.ts
   create src/commands/node/metadata.ts
   create src/commands/passphrase/decrypt.ts
   create src/commands/passphrase/encrypt.ts
   create src/commands/transaction/create.ts
   create src/commands/transaction/get.ts
   create src/commands/transaction/send.ts
   create src/commands/transaction/sign.ts
   create test/unit/modules/.gitkeep

>    create .gitignore

Author: Remove from lines 173-227?

=== Using the application CLI globally

To use the application commands globally, create an alias in the `.bashrc` or `.zshrc` file depending on the operating system that you are using.

[tabs]
====
bashrc::
+
--
.~/.bashrc/
[source,bash]
----
alias hello_app="$HOME/hello_app/bin/run"
----
Add the path to where your `hello_app` is located.

After updating the `.bashrc` file, make it directly available in the terminal by executing the following:

[source,bash]
----
. ~/.bashrc
----
--
zshrc::
+
--
.~/.zshrc/
[source,bash]
----
alias hello_app="$HOME/hello_app/bin/run" 
----
Add the path to where your `hello_app` is located.

After updating the `.zshrc` file, make it directly available in the terminal by executing the following:

[source,bash]
----
. ~/.zshrc
----
--
====

// .~/.bashrc/
// [source,bash]
// ----
// alias hello_app="$HOME/hello_app/bin/run" # <1>
// ----

// <1> Navigate to the path where your `hello_app` is located.

// After updating the `.bashrc` file, make it directly available in the terminal by executing the following:

// [source,bash]
// ----
// . ~/.bashrc

> // ----

Reviewer 2: imho this can be removed

.Available commands
----
Lisk-SDK Application

VERSION
  hello_app/0.1.0 darwin-arm64 node-v16.18.0

USAGE
  $ hello_app [COMMAND]

TOPICS
  block          Commands relating to hello_app blocks.
  blockchain     Commands relating to hello_app blockchain data.
  config         Commands relating to hello_app node configuration.
  endpoint       Commands relating to hello_app endpoint.
  generator      Commands relating to hello_app block generator.
  genesis-block  Creates a genesis block file.
  keys           Commands relating to hello_app key generation.
  node           Commands relating to hello_app node.
  passphrase     Commands relating to hello_app passphrases.
  transaction    Commands relating to hello_app transactions.

COMMANDS
  autocomplete  Displays autocomplete installation instructions
  console       Lisk interactive REPL session to run commands.
  hash-onion    Creates hash onions to be used by the forger.
  help          Displays help for hello_app.
  start         Starts Blockchain Node.
  version
----

> === Starting the application

Reviewer 2: ```suggestion
=== File- & folder structure
```

<3> `app/`: Contains the files of the blockchain application.
<4> `app.ts`: Creates the `Application` instance.
<5> `modules/`: Contains modules of the application.
This folder is empty after the first initialization with the `lisk init`.
The command `lisk generate:module` creates a new module in this folder.
<6> `modules.ts` Registers the modules with the application.
<7> `plugins/`: Contains plugins of the application.
This folder is empty after the first initialization with `lisk init`.
The command `lisk generate:plugin` creates a new plugin in this folder.
<8> `plugins.ts` Registers the plugins with the application.
<9> `commands/`: Contains the logic for the CLI commands of the application.
The files for the different commands can be adjusted and extended as desired, for example, to include new flags and commands.
<10> `test/`: Contains the test files for the unit, functional, and integration tests.

These files create a ready-to-start blockchain application configured for a local devnet, which uses only the default modules of the Lisk SDK.

> TIP: Lisk provides a range of default modules out of the box. These modules are created automatically whenever a blockchain application is bootstrapped via Lisk Commander. For more information, see the xref:{url_lisk_default_modules}[].

Reviewer: You can mention that they can simply update the line
`const { app } = Application.defaultApplication(config);` in app.ts to load application without default modules
```
	const app = new Application(config);

```

<3> `app/`: Contains the files of the blockchain application.
<4> `app.ts`: Creates the `Application` instance.
<5> `modules/`: Contains modules of the application.
This folder is empty after the first initialization with the `lisk init`.
The command `lisk generate:module` creates a new module in this folder.
<6> `modules.ts` Registers the modules with the application.
<7> `plugins/`: Contains plugins of the application.
This folder is empty after the first initialization with `lisk init`.
The command `lisk generate:plugin` creates a new plugin in this folder.
<8> `plugins.ts` Registers the plugins with the application.
<9> `commands/`: Contains the logic for the CLI commands of the application.
The files for the different commands can be adjusted and extended as desired, for example, to include new flags and commands.
<10> `test/`: Contains the test files for the unit, functional, and integration tests.

These files create a ready-to-start blockchain application configured for a local devnet, which uses only the default modules of the Lisk SDK.

> TIP: Lisk provides a range of default modules out of the box. These modules are created automatically whenever a blockchain application is bootstrapped via Lisk Commander. For more information, see the xref:{url_lisk_default_modules}[].

Reviewer 2: ```suggestion
[#default-modules]
.Default modules
TIP: Lisk provides a range of *default modules* out of the box.
These modules are created automatically, whenever a blockchain application is bootstrapped via Lisk Commander.
The default modules provide basic blockchain functionality that a most basic sidechain client requires to run.
For more information, see the xref:{url_lisk_default_modules}[].
```

> 

Reviewer 2: ```suggestion

=== Starting the application
```

The application is created in the file `app.ts`:

.src/app/app.ts
[source,typescript]
----
import { Application, PartialApplicationConfig } from 'lisk-sdk';
import { registerModules } from './modules';
import { registerPlugins } from './plugins';

export const getApplication = (config: PartialApplicationConfig): Application => {
	const { app } = Application.defaultApplication(config); //Creates a blockchain application with the default modules.
  
  // Will register additional modules to the application. 
  // Currently, no additional modules are available for the application.
  // To add new modules update the `modules.ts` file.
	registerModules(app);

  // Will register additional plugins to the application. 
  // Currently, no plugins are available for the application.
  // To add new plugins update the `plugins.ts` file.
	registerPlugins(app);

	return app;
};

----

> Alternatively, you can also run the application without the default modules by,
> replacing 

Reviewer 2: ```suggestion
In most cases, the <<default-modules,default modules>> don't need to be changed, as they provde basic funationality that most sidechain clients need.
But in case you want to replace the default modules with other modules, you can also run the application without the default modules by replacing 
```

[source,typescript]
----
const { app } = Application.defaultApplication(config);
----
with

[source,typescript]
----
const app = new Application(config);

> ----

Reviewer 2: ```suggestion
----

Please be aware that if you create the app with `const app = new Application(config);`, then you need to register all modules manually in the `modules.ts` file.

This way it is also possible to include some, but not all of the default modules in the client application.
```

import { registerModules } from './modules';
import { registerPlugins } from './plugins';

export const getApplication = (config: PartialApplicationConfig): Application => {
	const { app } = Application.defaultApplication(config); //Creates a blockchain application with the default modules.
  
  // Will register additional modules to the application. 
  // Currently, no additional modules are available for the application.
  // To add new modules update the `modules.ts` file.
	registerModules(app);

  // Will register additional plugins to the application. 
  // Currently, no plugins are available for the application.
  // To add new plugins update the `plugins.ts` file.
	registerPlugins(app);

	return app;
};

----

In most cases, the <<default-modules,default modules>> don't need to be changed, as they provide the basic functionality that most sidechain clients need.
But in case you want to replace the default modules with other modules, you can also run the application without the default modules by replacing 

[source,typescript]
----
const { app } = Application.defaultApplication(config);
----
with

[source,typescript]
----
const app = new Application(config);
----

> Please be aware that if you create the app with `const app = new Application(config);`, then you need to register all modules manually in the `modules.ts` file.

Reviewer: ```suggestion
Please be aware that if you create the app with `const app = new Application(config);`, then you need to register all modules manually in the `app.ts` file.
```

│   │   ├── plugins/ <7>
│   │   └── plugins.ts <8>
│   └── commands/ <9>
├── test/ <10>
├── jest.config.js
├── package-lock.json
├── package.json
├── readme.md
└── tsconfig.json
----

<1> `bin/`: Contains the script to run the CLI of the application.
<2> `config/`: Contains the configuration, dev-validators, genesis assets, genesis block, and passphrase files that are used by the application.
<3> `app/`: Contains the files of the blockchain application.
<4> `app.ts`: Creates the `Application` instance.

> <5> `modules/`: Contains modules of the application.

Reviewer 2: regarding my comment below, i think it is important to not that it only include the modules defined by the developer, not the default modules of the client.

This folder is empty after the first initialization with the `lisk init`.
The command `lisk generate:module` creates a new module in this folder.
<6> `modules.ts` Registers the modules with the application.
<7> `plugins/`: Contains plugins of the application.
This folder is empty after the first initialization with `lisk init`.
The command `lisk generate:plugin` creates a new plugin in this folder.
<8> `plugins.ts` Registers the plugins with the application.
<9> `commands/`: Contains the logic for the CLI commands of the application.
The files for the different commands can be adjusted and extended as desired, for example, to include new flags and commands.
<10> `test/`: Contains the test files for the unit, functional, and integration tests.

> These files create a ready-to-start blockchain application configured for a local devnet, which uses only the default modules of the Lisk SDK.

Reviewer 2: I think we didn't introduce to the reader yet, what are the modules which are part of the client by default? We should introduce the term and meaning of `default modules`, probably on page `understand-blockchain/index.adoc`. At this place, we could then link to that section explaining the default modules.

.src/app/app.ts
[source,typescript]
----
import { Application, PartialApplicationConfig } from 'lisk-sdk';
import { registerModules } from './modules';
import { registerPlugins } from './plugins';

export const getApplication = (config: PartialApplicationConfig): Application => {
	const { app } = Application.defaultApplication(config); <1>

	registerModules(app); <2>

> 	registerPlugins(app); <3>

Reviewer 2: Before this issue is fixed, we shouldn't use callouts in the code snippets: https://github.com/LiskHQ/lisk-docs/issues/1481
